core/org.wso2.carbon.core.services/src/main/java/org/wso2/carbon/core/services/fileupload/FileUploadService.java

```diff
@@ -79,15 +83,42 @@
         }
     }
 
+    private void verifyCanonicalDestination(String extraFileLocation, File dirs, String fileName) throws IOException {
+
+        String canonicalDestinationDirPath = dirs.getCanonicalPath();
+        File destinationFile = new File(extraFileLocation, fileName);
+        String canonicalDestinationFile = destinationFile.getCanonicalPath();
+
+        if (!canonicalDestinationFile.startsWith(canonicalDestinationDirPath + File.separator)) {
+            throw new AxisFault("Entry is outside of the target dir: " + fileName);
+        }
+    }
+
     private String getFileName(String fileName) {
-        String fileNameOnly;
-        if (fileName.indexOf("\\") < 0) {
-            fileNameOnly = fileName.substring(fileName.lastIndexOf('/') + 1,
-                                              fileName.length());
-        } else {
-            fileNameOnly = fileName.substring(fileName.lastIndexOf("\\") + 1,
-                                              fileName.length());
+
+        if (fileName == null) {
+            return null;
+        }
+        return requireNonNullChars(fileName).substring(indexOfLastSeparator(fileName) + 1);
+    }
+
+    private int indexOfLastSeparator(final String fileName) {
+
+        if (fileName == null) {
+            return -1;
+        }
+        final int lastUnixPos = fileName.lastIndexOf(UNIX_SEPARATOR);
```

Comment:
Java should have a method or variable to resolve the OS-specific separator.

Reply:
We need input from security team regarding this. As I feel this validation is to avoid manipulating fileName by using / and \\. We cannot achieve this by using OS-specific separators.

Reply:
It's true that we can use [https://docs.oracle.com/javase/8/docs/api/java/io/File.html#pathSeparator](https://docs.oracle.com/javase/8/docs/api/java/io/File.html#pathSeparator). However, this will change the behaviour of the method. It's a risk we should not take at this time. The method is designed to return the filename despite the type of path (windows / linux) provided to the method. By reusing [https://commons.apache.org/proper/commons-io/javadocs/api-1.4/org/apache/commons/io/FilenameUtils.html#getName(java.lang.String)](https://commons.apache.org/proper/commons-io/javadocs/api-1.4/org/apache/commons/io/FilenameUtils.html#getName(java.lang.String)) we are preserving same behaviour.